examples/perceive_and_act/custom_agents.py

```diff
@@ -0,0 +1,79 @@
+
+class PerceivingAgent(WiseAgent):
+    yaml_tag = u'!custom_agents.PerceivingAgent'
+    yaml_loader = WiseAgentsLoader
+
+
+    def __init__(self, name: str, description: str, transport: WiseAgentTransport, file_path: str, check_interval: float, destination_agent_name: str):
+        self._file_path = file_path
+        self._check_interval = check_interval
+        self._destination_agent_name = destination_agent_name
+        super().__init__(name=name, description=description, transport=transport)
+
+    def start_agent(self):
+        super().start_agent()
+        self.perceive(self._file_path, self.on_file_change, self._check_interval)
+        
+    def perceive(self, file_path, callback, check_interval=1.0):
+        """
+        Monitors a file for any changes and invokes a callback with the file's content upon modification.
+
+        
+        Args:
+            file_path (str): The path to the file to monitor.
+            callback (callable): A function to call with the file's content when it changes.
+            check_interval (float, optional): Time in seconds between checks. Defaults to 1.0.
+        """
+        def watch():
+            try:
+                last_mtime = os.path.getmtime(file_path)
+            except FileNotFoundError:
+                last_mtime = None
+
+            while True:
+                try:
+                    print(f"Checking file {file_path}")
+                    current_mtime = os.path.getmtime(file_path)
+                    if last_mtime is None:
+                        last_mtime = current_mtime
+                    elif current_mtime != last_mtime:
+                        print(f"File {file_path} has changed")
+                        last_mtime = current_mtime
+                        with open(file_path, 'r') as f:
+                            content = f.read()
+                        callback(content)
+                except FileNotFoundError:
+                    if last_mtime is not None:
+                        last_mtime = None
+                except Exception as e:
+                    print(f"Error monitoring file: {e}")
+                time.sleep(check_interval)
+
+        thread = threading.Thread(target=watch, daemon=True)
+        thread.start()
+
+    def on_file_change(self, content):
+        print(f"sending message: {content}, {self.name}, {self._destination_agent_name}")
+        self.send_request(WiseAgentMessage(content, self.name), self._destination_agent_name)
+
+class ActionAgent(WiseAgent):
+    yaml_tag = u'!custom_agents.ActionAgent'
+    yaml_loader = WiseAgentsLoader
+
+    def __init__(self, name: str, description: str, transport: WiseAgentTransport, destination_file_path: str):
+        self._destination_file_path = destination_file_path
+        super().__init__(name=name, description=description, transport=transport)
+
+    def start_agent(self):
+        super().start_agent()
+
+    def handle_request(self, message: WiseAgentMessage):
```

Comment:
Instead of handle_request, the process_request method should now be implemented as described here:

https://github.com/wise-agents/wise-agents/blob/main/docs/custom_agents.md

(handle_request is no longer abstract)